### PR TITLE
XMLBear: Process error correctly

### PR DESCRIPTION
**Pull request by vivek425ster**
Errors were shown as logs instead of errors. Corrected the processing of
output to show the error messages.

Fixes https://github.com/coala-analyzer/coala-bears/issues/251

### CODE REVIEW

**Comment by sils1297** [comment]
cancelling non circle builds here as this needs to be rebased later anyway

**Comment by sils1297** [comment]
repeat: the builds are not failed :) all is well

File: bears/xml/XMLBear.py [hunk 1]
```diff
@@ -24,10 +24,11 @@ class XMLBear(LocalBear, Lint):
     gives_corrected = True
 
     def process_output(self, output, filename, file):
-        if self.stdout_output:  # only yield Result if stdout is not empty
-            return self._process_corrected(self.stdout_output, filename, file)
-        if self.stderr_output:  # pragma: no cover
+        if self.stderr_output:
+            self.use_stderr = True
```
Comment (sils1297) [inline bears/xml/XMLBear.py]: just define it outside, like gives corrected

**Comment by sils1297** [comment]
unack f7fa889

File: bears/xml/XMLBear.py [hunk 2]
```diff
@@ -24,10 +24,11 @@ class XMLBear(LocalBear, Lint):
     gives_corrected = True
 
     def process_output(self, output, filename, file):
-        if self.stdout_output:  # only yield Result if stdout is not empty
-            return self._process_corrected(self.stdout_output, filename, file)
-        if self.stderr_output:  # pragma: no cover
+        self.use_stderr = True
```
Comment (sils1297) [inline bears/xml/XMLBear.py]: this needs to be moved to class level

File: bears/xml/XMLBear.py [hunk 3]
```diff
@@ -22,12 +22,13 @@ class XMLBear(LocalBear, Lint):
...
+    use_stderr = True
 
     def process_output(self, output, filename, file):
-        if self.stdout_output:  # only yield Result if stdout is not empty
-            return self._process_corrected(self.stdout_output, filename, file)
-        if self.stderr_output:  # pragma: no cover
+        if self.stderr_output:
             return self._process_issues(self.stderr_output, filename)
+        if self.stdout_output:
+            return self._process_corrected(self.stdout_output, filename, file)
```
Comment (sils1297) [inline bears/xml/XMLBear.py]: why do you switch those two and why do you remove the comment in line 27?

Comment (vivek425ster) [inline bears/xml/XMLBear.py]: @sils1297 I switched it because if there is an error in stderr then return no need to check stdout else check if something needs to be corrected

Comment (AbdealiJK) [inline bears/xml/XMLBear.py]: If this is a bug that you're trying to fix, please make a new test to test this.

Either way it would be good to add a comment here stating this so that someone else doesnt change it later by mistake
Note: You can use `elif` instead of `if` here.

Comment (vivek425ster) [inline bears/xml/XMLBear.py]: @AbdealiJK Tests are already there which covers everything.
What more tests you want me to add ?

**Comment by vivek425ster** [comment]
@SanketDG @AbdealiJK Any particular reason why coverage is not 100%. 
From what i know https://github.com/coala-analyzer/coala-bears/blob/master/tests/xml/XMLBearTest.py#L11 this test should cover the elif case .

File: bears/xml/XMLBear.py [hunk 4]
```diff
@@ -24,10 +26,16 @@ class XMLBear(LocalBear, Lint):
...
 
     def process_output(self, output, filename, file):
-        if self.stdout_output:  # only yield Result if stdout is not empty
-            return self._process_corrected(self.stdout_output, filename, file)
-        if self.stderr_output:  # pragma: no cover
+        if self.stderr_output:
+            self.use_stderr = True
+            if self.stdout_output:  # only yield Result if stdout is not empty
+                return collections.ChainMap(
+                           self._process_issues(self.stderr_output, filename),
```
Comment (AbdealiJK) [inline bears/xml/XMLBear.py]: indentation ?

File: bears/xml/XMLBear.py [hunk 5]
```diff
@@ -24,10 +26,16 @@ class XMLBear(LocalBear, Lint):
     gives_corrected = True
 
     def process_output(self, output, filename, file):
-        if self.stdout_output:  # only yield Result if stdout is not empty
-            return self._process_corrected(self.stdout_output, filename, file)
-        if self.stderr_output:  # pragma: no cover
+        if self.stderr_output:
+            self.use_stderr = True
+            if self.stdout_output:  # only yield Result if stdout is not empty
+                return collections.ChainMap(
```
Comment (AbdealiJK) [inline bears/xml/XMLBear.py]: I don't understand why chainMap is needed. This is a list of Results.
chainMap is for dicts right ?

File: bears/xml/XMLBear.py [hunk 6]
```diff
@@ -24,10 +26,16 @@ class XMLBear(LocalBear, Lint):
     gives_corrected = True
 
     def process_output(self, output, filename, file):
-        if self.stdout_output:  # only yield Result if stdout is not empty
-            return self._process_corrected(self.stdout_output, filename, file)
-        if self.stderr_output:  # pragma: no cover
+        if self.stderr_output:
+            self.use_stderr = True
+            if self.stdout_output:  # only yield Result if stdout is not empty
```
Comment (AbdealiJK) [inline bears/xml/XMLBear.py]: I think the comment doesnt explain things correctly ?
We yield results even if stdout is empty at L36

File: bears/xml/XMLBear.py (hunk 1 above)
Comment (AbdealiJK) [inline bears/xml/XMLBear.py]: use_stderr isnt really needed. It is used only in `lint()` and wont be used after this line changes it

File: bears/xml/XMLBear.py [hunk 7]
```diff
@@ -24,10 +26,16 @@ class XMLBear(LocalBear, Lint):
...
+                return collections.ChainMap(
+                           self._process_issues(self.stderr_output, filename),
+                           self._process_corrected(self.stdout_output,
+                                                   filename, file))
             return self._process_issues(self.stderr_output, filename)
+        elif self.stdout_output:  # only yield Result if stdout is not empty
```
Comment (AbdealiJK) [inline bears/xml/XMLBear.py]: Again, this comment isnt helping - Im not sure what youre trying to say. We're yielding Results everywhere right ?

File: bears/xml/XMLBear.py (hunk 1 above)
Comment (vivek425ster) [inline bears/xml/XMLBear.py]: @AbdealiJK use_stderr is needed because if not set to True it will print logs also

Comment (AbdealiJK) [inline bears/xml/XMLBear.py]: Ah. you're right. Sorry - I got confused.
@sils1297 it cannot be moved outside.

File: bears/xml/XMLBear.py (hunk 4 above)
Comment (vivek425ster) [inline bears/xml/XMLBear.py]: @AbdealiJK What's the issue in indentation ?
Isn't it 4 spaces PEP8 doesn't give me any error.

File: bears/xml/XMLBear.py (hunk 5 above)
Comment (vivek425ster) [inline bears/xml/XMLBear.py]: @AbdealiJK yeah ``chainMap`` is for dict. 
Changing it to ``itertools.chain``

File: bears/xml/XMLBear.py (hunk 4 above)
Comment (AbdealiJK) [inline bears/xml/XMLBear.py]: It should be 4 spaces from the beginning of "return"

File: bears/xml/XMLBear.py (hunk 6 above)
Comment (vivek425ster) [inline bears/xml/XMLBear.py]: How does ``only merge corrections if stdout is not empty`` sound ?

File: bears/xml/XMLBear.py (hunk 7 above)
Comment (vivek425ster) [inline bears/xml/XMLBear.py]: @AbdealiJK It is specific to that ``elif`` clause. Won't it yield result only when there is some correction to be made ?

File: bears/xml/XMLBear.py (hunk 6 above)
Comment (AbdealiJK) [inline bears/xml/XMLBear.py]: What about "Return issues from stderr and diffs from stdout if both are not empty"

**Comment by vivek425ster** [comment]
@SanketDG @AbdealiJK Why isn't this elif getting covered ?
https://codecov.io/github/coala-analyzer/coala-bears/bears/xml/XMLBear.py?ref=f0425f051fba52543fd2984750a8277b427e655a#l-37
From what i know https://github.com/coala-analyzer/coala-bears/blob/master/tests/xml/XMLBearTest.py#L11 this test should cover the elif case .

File: bears/xml/XMLBear.py (hunk 6 above)
Comment (vivek425ster) [inline bears/xml/XMLBear.py]: :+1:

File: bears/xml/XMLBear.py [hunk 8]
```diff
@@ -24,10 +26,11 @@ class XMLBear(LocalBear, Lint):
     gives_corrected = True
 
     def process_output(self, output, filename, file):
-        if self.stdout_output:  # only yield Result if stdout is not empty
-            return self._process_corrected(self.stdout_output, filename, file)
-        if self.stderr_output:  # pragma: no cover
-            return self._process_issues(self.stderr_output, filename)
+        if self.stderr_output:
+            self.use_stderr = True
```
Comment (SanketDG) [inline bears/xml/XMLBear.py]: not sure if this is needed, why do you keep this here?
The usage of `use_stderr` is determined in the `Lint.lint()` function.

Comment (vivek425ster) [inline bears/xml/XMLBear.py]: @SanketDG It needs to be added because if it is not there then logs will also be printed

Comment (SanketDG) [inline bears/xml/XMLBear.py]: Then why not just `use_stderr=True`

File: bears/xml/XMLBear.py [hunk 9]
```diff
@@ -24,10 +26,11 @@ class XMLBear(LocalBear, Lint):
...
+        return itertools.chain(
+            self._process_issues(self.stderr_output, filename),
+            self._process_corrected(self.stdout_output, filename, file))
```
Comment (SanketDG) [inline bears/xml/XMLBear.py]: This won't work, because you are not checking if `self.stdout_output` is empty or not. Thus if I have:

`test.xml` as:

```xml
<a>hey & hi</a>
```

and I do:

```
❯ coala --files=test.xml --bears=XMLBear -L DEBUG
```

I get:

```
test.xml
|   1| <a>hey•&•hi</a>
|    | [NORMAL] XMLBear:
|    | XML can be formatted better.
|    | The following actions are applicable to this result:
|    |  0: Apply no further actions.
|    |  1: Open the affected file(s) in an editor.
|    |  2: Apply the patch automatically.
|    |  3: Print a diff of the patch that would be applied.
|    | Please enter the number of the action you want to execute. 3
|----|    | /home/sanket/coala-bears/test.xml
|    |++++| /home/sanket/coala-bears/test.xml
|   1|    |-<a>hey & hi</a>
|    | Displayed patch successfully.
|    | The following actions are applicable to this result:
|    |  0: Apply no further actions.
|    |  1: Open the affected file(s) in an editor.
|    |  2: Apply the patch automatically.
|    |  3: Print a diff of the patch that would be applied.
|    | Please enter the number of the action you want to execute. 
```

As you can see, the diff is removing the line, because `stdout` is empty.

**Comment by SanketDG** [comment]
unack 6d6e9e8

File: bears/xml/XMLBear.py (hunk 9 above)
Comment (vivek425ster) [inline bears/xml/XMLBear.py]: @SanketDG  I thought it gives stdout also when it gives error .
Because when you validate it against it a schema if invalid it gives output in stdout as well and also errors in stderr
```
<?xml version="1.0"?>
<note xmlns="http://www.w3schools.com" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="note.xsd">
   <from>Jani</from>
   <heading>Reminder</heading>
   <body>Don't forget me this weekend!</body>
</note>
note-invalid.xml:2: element from: Schemas validity error : Element '{http://www.w3schools.com}from': This element is not expected. Expected is ( {http://www.w3schools.com}to ).
note-invalid.xml fails to validate
```

File: bears/xml/XMLBear.py (hunk 8 above)
Comment (vivek425ster) [inline bears/xml/XMLBear.py]: @SanketDG if i do `use_stderr = True` without checking whether the stderr_output is there or not it will set ``results_output `` to stderr_output even when there are no any errors in stderr. 
https://github.com/coala-analyzer/coala/blob/master/coalib/bearlib/abstractions/Lint.py#L98

**Comment by vivek425ster** [comment]
@SanketDG @AbdealiJK @sils1297 Any idea why elif is not getting covered .From what i know https://github.com/coala-analyzer/coala-bears/blob/master/tests/xml/XMLBearTest.py#L11 this test should cover the elif case .
This test should cover the elif case.

**Comment by sils1297** [comment]
seems to be covered now?

**Comment by vivek425ster** [comment]
@sils1297 Yeah i got the problem it was because if both stdout and stderr were empty nothing was being done

File: tests/xml/test_files/xsd-error.xml [hunk 10]
```diff
@@ -1,4 +1,3 @@
-<?xml version="1.0"?>
```
Comment (Adrianzatreanu) [inline tests/xml/test_files/xsd-error.xml]: why is this needed to be removed?

Comment (AbdealiJK) [inline tests/xml/test_files/xsd-error.xml]: If I understand correctly, the stdout (which gives corrected output) adds it to the corrected one.
Hence, for coverage of Line 36.

But I havent followed this very closely. So, would be good to let @vivek425ster confirm

Comment (vivek425ster) [inline tests/xml/test_files/xsd-error.xml]: @Adrianzatreanu yes it is so that stdout gives corrections

File: bears/xml/XMLBear.py [hunk 11]
```diff
@@ -22,11 +24,16 @@ class XMLBear(LocalBear, Lint):
...
     def process_output(self, output, filename, file):
-        if self.stdout_output:  # only yield Result if stdout is not empty
-            return self._process_corrected(self.stdout_output, filename, file)
-        if self.stderr_output:  # pragma: no cover
+        if self.stdout_output:
+            # Return issues from stderr and stdout if stdout is not empty
+            return itertools.chain(
+                self._process_issues(self.stderr_output, filename),
+                self._process_corrected(
+                    self.stdout_output, filename, file))
```
Comment (SanketDG) [inline bears/xml/XMLBear.py]: why wrap this to the next line?

**Comment by SanketDG** [comment]
cf51a65 needs work

**Comment by SanketDG** [comment]
ack 06b5761

**Comment by SanketDG** [comment]
ack 06b5761

**Comment by sils1297** [comment]
@rultor merge
